extensions/amp-video-iframe/0.1/amp-video-iframe.js

```diff
@@ -0,0 +1,382 @@
+ * @private @const
+ */
+const ALLOWED_EVENTS = [
+  VideoEvents.PLAYING,
+  VideoEvents.PAUSE,
+  VideoEvents.ENDED,
+  VideoEvents.MUTED,
+  VideoEvents.UNMUTED,
+  VideoEvents.AD_START,
+  VideoEvents.AD_END,
+];
+
+/** @implements {../../../src/video-interface.VideoInterface} */
+class AmpVideoIframe extends AMP.BaseElement {
+
+  /** @param {!AmpElement} element */
+  constructor(element) {
+    super(element);
+
+    /** @private {boolean} */
+    this.isAdLike_ = false;
+
+    /** @private {boolean} */
+    this.isTrackingIframe_ = false;
+
+    /** @private {?Element} */
+    this.iframe_ = null;
+
+    /** @private {!UnlistenDef|null} */
+    this.unlistenFrame_ = null;
+
+    /** @private {?function()} */
+    this.readyPromise_ = null;
+
+    /** @private {?function()} */
+    this.readyResolver_ = null;
+
+    /** @private {?function()} */
+    this.readyRejecter_ = null;
+
+    /** @private {boolean} */
+    this.canPlay_ = false;
+
+    /**
+     * @param {!Event} e
+     * @private
+     */
+    this.boundOnMessage_ = e => this.onMessage_(e);
+  }
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+
+  /** @override */
+  onLayoutMeasure() {
+    const {element} = this;
+    this.isAdLike_ = isAdLike(element);
+    this.isTrackingIframe_ = looksLikeTrackingIframe(element.getLayoutBox());
+  }
+
+  /** @override */
+  getLayoutPriority() {
+    if (this.isAdLike_) {
+      return LayoutPriority.ADS; // See AmpAd3PImpl.
+    }
+    if (this.isTrackingIframe_) {
+      return LayoutPriority.METADATA;
+    }
+    return super.getLayoutPriority();
+  }
+
+  /** @override */
+  buildCallback() {
+    installVideoManagerForDoc(this.getAmpDoc());
+  }
+
+  /** @override */
+  layoutCallback() {
+    this.iframe_ = createFrameFor(this, this.getSrc_(), SANDBOX);
+    this.unlistenFrame_ = listen(this.win, 'message', this.boundOnMessage_);
+    return this.createReadyPromise_().then(() => this.onReady_());
+  }
+
+  /** @private */
+  onReady_() {
+    const {element} = this;
+    Services.videoManagerForDoc(element).register(this);
+    element.dispatchCustomEvent(VideoEvents.LOAD);
+  }
+
+  /** @override */
+  createPlaceholderCallback() {
+    const {element} = this;
+    const poster =
+        htmlFor(element)`<amp-img layout=fill placeholder></amp-img>`;
+
+    poster.setAttribute('src',
+        this.user().assertString(element.getAttribute('poster')));
+
+    return poster;
+  }
+
+  /** @override */
+  unlayoutCallback() {
+    this.removeIframe_();
+    return true; // layout again.
+  }
+
+  /** @private */
+  removeIframe_() {
+    if (this.iframe_) {
+      removeElement(this.iframe_);
+      this.iframe_ = null;
+    }
+    if (this.unlistenFrame_) {
+      this.unlistenFrame_();
+      this.unlistenFrame_ = null;
+      this.canPlay_ = false;
+    }
+  }
+
+  /** @private */
+  getSrc_() {
+    // TODO: assert https
+    return this.element.getAttribute('src');
```

Comment:
we should check if origin of src is same as page and show a warning such as "Origin of the video-iframe and the page are the same which allows for same-origin behavior, however in AMP cache, origins won't match. Please ensure you do not rely on any same-origin privileges."

`amp-iframe` currently throws on this but technically this is not a security thing and we want to relax it anyway: https://github.com/ampproject/amphtml/issues/8545 so we can start by relaxing it here.